File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/data/notes/NoteService.kt
package com.arcinteractive.spaces.data.notes

import android.content.Context
import android.util.Base64
import com.arcinteractive.spaces.data.model.*
import com.arcinteractive.spaces.data.spaces.EncryptionService
import com.arcinteractive.spaces.data.spaces.SpaceService
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.*
import com.google.firebase.storage.FirebaseStorage
import java.util.Date
import java.util.UUID
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine
import org.json.JSONArray
import org.json.JSONObject

class NoteService {
    private val firestore = FirebaseFirestore.getInstance()
    private val storage = FirebaseStorage.getInstance()
    private val encryption = EncryptionService()
    private val spaces = SpaceService()

    fun listen(context: Context, space: Space, onUpdate: (Result<List<SpaceNote>>) -> Unit): ListenerRegistration =
        firestore.collection("spaces").document(space.id).collection("notes").orderBy("updatedAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) return@addSnapshotListener onUpdate(Result.failure(error))
                val docs = snapshot?.documents.orEmpty()
                if (docs.isEmpty()) return@addSnapshotListener onUpdate(Result.success(emptyList()))
                Thread {
                    runCatching {
                        val key = kotlinx.coroutines.runBlocking { spaces.encryptionKeyForModuleData(context, space.id) }
                        docs.map { mapNote(it, space.id, key) }
                    }.also(onUpdate)
                }.start()
            }

    fun listenComments(context: Context, space: Space, noteId: String, onUpdate: (Result<List<SpaceNoteComment>>) -> Unit): ListenerRegistration =
        firestore.collection("spaces").document(space.id).collection("notes").document(noteId).collection("comments").orderBy("createdAt")
            .addSnapshotListener { snapshot, error ->
                if (error != null) return@addSnapshotListener onUpdate(Result.failure(error))
                val docs = snapshot?.documents.orEmpty()
                if (docs.isEmpty()) return@addSnapshotListener onUpdate(Result.success(emptyList()))
                Thread {
                    runCatching {
                        val key = kotlinx.coroutines.runBlocking { spaces.encryptionKeyForModuleData(context, space.id) }
                        docs.map { doc ->
                            val json = JSONObject(encryption.decryptText(doc.getString("ciphertext").orEmpty(), doc.getString("nonce").orEmpty(), null, key))
                            SpaceNoteComment(doc.id, noteId, doc.getString("authorId").orEmpty(), json.getString("authorName"), json.getString("body"), doc.getTimestamp("createdAt")?.toDate() ?: Date())
                        }
                    }.also(onUpdate)
                }.start()
            }

    suspend fun save(context: Context, space: Space, note: SpaceNote) {
        val uid = FirebaseAuth.getInstance().currentUser?.uid ?: error("Sign in to use Notes.")
        val isNew = note.createdBy.isBlank()
        val permission = if (isNew) SpacePermission.CreateNotes else if (note.createdBy == uid) SpacePermission.EditOwnNotes else SpacePermission.EditAnyNotes
        check(spaces.canPerform(context, space, permission)) { "You do not have permission to change this Note." }
        val json = JSONObject().put("title", note.title).put("markdown", note.markdown)
            .put("attachments", JSONArray(note.attachments.map(::attachmentJson)))
            .put("links", JSONArray(note.links.map(::linkJson)))
        val encrypted = encryption.encryptText(json.toString(), spaces.encryptionKeyForModuleData(context, space.id))
        awaitSet(firestore.collection("spaces").document(space.id).collection("notes").document(note.id), mapOf(
            "ciphertext" to encrypted.ciphertext, "nonce" to encrypted.nonce,
            "createdBy" to if (isNew) uid else note.createdBy,
            "createdAt" to if (isNew) FieldValue.serverTimestamp() else note.createdAt,
            "updatedAt" to FieldValue.serverTimestamp(), "encryptionVersion" to "space-module-key-v1"
        ))
        if (isNew) {
            spaces.recordModuleActivity(
                context, ActivityType.NoteCreated, space, "created a Note", note.title,
                note.id, ActivityTargetType.Notes
            )
        }
        if (!isNew) {
            val recipients = spaces.mentionedMemberIds(context, note.markdown, space)
            spaces.recordTargetedModuleNotification(
                context, recipients, "note", space, "mentioned you in a Note", note.title,
                note.id, ActivityTargetType.Notes
            )
        }
    }

    suspend fun delete(context: Context, space: Space, note: SpaceNote) {
        val uid = FirebaseAuth.getInstance().currentUser?.uid
        check(spaces.canPerform(context, space, if (note.createdBy == uid) SpacePermission.DeleteOwnNotes else SpacePermission.DeleteAnyNotes))
        note.attachments.forEach { runCatching { deleteTrackedStorage(context, space.id, it.storagePath) } }
        val ref = firestore.collection("spaces").document(space.id).collection("notes").document(note.id)
        awaitGet(ref.collection("comments")).documents.forEach { awaitDelete(it.reference) }
        awaitDelete(ref)
    }

    suspend fun addComment(context: Context, space: Space, note: SpaceNote, body: String, authorName: String) {
        val uid = FirebaseAuth.getInstance().currentUser?.uid ?: error("Sign in to comment.")
        val json = JSONObject().put("authorName", authorName).put("body", body)
        val encrypted = encryption.encryptText(json.toString(), spaces.encryptionKeyForModuleData(context, space.id))
        awaitSet(firestore.collection("spaces").document(space.id).collection("notes").document(note.id).collection("comments").document(), mapOf(
            "ciphertext" to encrypted.ciphertext, "nonce" to encrypted.nonce, "authorId" to uid,
            "createdAt" to FieldValue.serverTimestamp(), "encryptionVersion" to "space-module-key-v1"
        ))
        val commentRecipients = (spaces.mentionedMemberIds(context, body, space) + note.createdBy).distinct()
        spaces.recordTargetedModuleNotification(
            context, commentRecipients, "note", space, "commented on your Note", note.title,
            note.id, ActivityTargetType.Notes
        )
    }

    suspend fun recordView(spaceId: String, noteId: String) {
        val uid = FirebaseAuth.getInstance().currentUser?.uid ?: return
        awaitSet(firestore.collection("spaces").document(spaceId).collection("notePreferences").document(uid).collection("notes").document(noteId), mapOf(
            "lastViewedAt" to FieldValue.serverTimestamp(), "viewCount" to FieldValue.increment(1)
        ))
    }

    fun listenViewPreferences(spaceId: String, onUpdate: (Map<String, NoteViewPreference>) -> Unit): ListenerRegistration? {
        val uid = FirebaseAuth.getInstance().currentUser?.uid ?: return null
        return firestore.collection("spaces").document(spaceId).collection("notePreferences")
            .document(uid).collection("notes").addSnapshotListener { snapshot, _ ->
                onUpdate(snapshot?.documents.orEmpty().associate { document ->
                    document.id to NoteViewPreference(
                        document.getTimestamp("lastViewedAt")?.toDate() ?: Date(0),
                        document.getLong("viewCount")?.toInt() ?: 0
                    )
                })
            }
    }

    fun listenManualOrder(spaceId: String, onUpdate: (List<String>) -> Unit): ListenerRegistration? {
        val uid = FirebaseAuth.getInstance().currentUser?.uid ?: return null
        return firestore.collection("spaces").document(spaceId).collection("notePreferences")
            .document(uid).addSnapshotListener { snapshot, _ ->
                onUpdate((snapshot?.get("manualOrder") as? List<*>)?.filterIsInstance<String>().orEmpty())
            }
    }

    suspend fun saveManualOrder(spaceId: String, ids: List<String>) {
        val uid = FirebaseAuth.getInstance().currentUser?.uid ?: return
        awaitSet(
            firestore.collection("spaces").document(spaceId).collection("notePreferences").document(uid),
            mapOf("manualOrder" to ids, "updatedAt" to FieldValue.serverTimestamp())
        )
    }

    suspend fun upload(context: Context, space: Space, noteId: String, bytes: ByteArray, name: String, mimeType: String, isMedia: Boolean): SpaceNoteAttachment {
        require(bytes.size.toLong() <= 1024L * 1024L * 1024L) { "Each upload is limited to 1 GB." }
        val organizations = com.arcinteractive.spaces.data.organization.OrganizationService()
        val organizationId = organizations.reserveStorage(context, space.id, bytes.size.toLong())
        val id = UUID.randomUUID().toString()
        val encrypted = encryption.encryptBytes(bytes, spaces.encryptionKeyForModuleData(context, space.id))
        val path = "spaces/${space.id}/notes/$noteId/attachments/$id.enc"
        try {
            awaitUpload(path, Base64.decode(encrypted.ciphertext, Base64.NO_WRAP), bytes.size.toLong())
        } catch (error: Throwable) {
            organizations.releaseStorage(context, organizationId, bytes.size.toLong())
            throw error
        }
        return SpaceNoteAttachment(id, name, mimeType, path, encrypted.nonce, isMedia)
    }

    suspend fun download(context: Context, space: Space, attachment: SpaceNoteAttachment): ByteArray {
        val bytes = awaitDownload(attachment.storagePath)
        return encryption.decryptBytes(Base64.encodeToString(bytes, Base64.NO_WRAP), attachment.nonce, spaces.encryptionKeyForModuleData(context, space.id))
    }

    private fun mapNote(doc: DocumentSnapshot, spaceId: String, key: ByteArray): SpaceNote {
        val json = JSONObject(encryption.decryptText(doc.getString("ciphertext").orEmpty(), doc.getString("nonce").orEmpty(), null, key))
        return SpaceNote(doc.id, spaceId, json.getString("title"), json.optString("markdown"),
            json.optJSONArray("attachments").objects().map(::mapAttachment), json.optJSONArray("links").objects().map(::mapLink),
            doc.getString("createdBy").orEmpty(), doc.getTimestamp("createdAt")?.toDate() ?: Date(), doc.getTimestamp("updatedAt")?.toDate() ?: Date())
    }
    private fun linkJson(v: SpaceLinkAttachment) = JSONObject().put("id", v.id).put("moduleType", v.moduleType.name).put("targetId", v.targetId).put("title", v.title).put("subtitle", v.subtitle).put("icon", v.icon)
    private fun mapLink(v: JSONObject) = SpaceLinkAttachment(v.getString("id"), SpaceLinkModuleType.valueOf(v.getString("moduleType")), v.getString("targetId"), v.getString("title"), v.optString("subtitle").ifBlank { null }, v.optString("icon"))
    private fun attachmentJson(v: SpaceNoteAttachment) = JSONObject().put("id", v.id).put("name", v.name).put("mimeType", v.mimeType).put("storagePath", v.storagePath).put("nonce", v.nonce).put("isMedia", v.isMedia)
    private fun mapAttachment(v: JSONObject) = SpaceNoteAttachment(v.getString("id"), v.getString("name"), v.getString("mimeType"), v.getString("storagePath"), v.getString("nonce"), v.optBoolean("isMedia"))
    private fun JSONArray?.objects() = (0 until (this?.length() ?: 0)).map { this!!.getJSONObject(it) }
    private suspend fun awaitSet(ref: DocumentReference, data: Map<String, Any>) = suspendCancellableCoroutine<Unit> { c -> ref.set(data, SetOptions.merge()).addOnSuccessListener { c.resume(Unit) {} }.addOnFailureListener(c::resumeWithException) }
    private suspend fun awaitDelete(ref: DocumentReference) = suspendCancellableCoroutine<Unit> { c -> ref.delete().addOnSuccessListener { c.resume(Unit) {} }.addOnFailureListener(c::resumeWithException) }
    private suspend fun awaitGet(ref: CollectionReference) = suspendCancellableCoroutine<QuerySnapshot> { c -> ref.get().addOnSuccessListener { c.resume(it) {} }.addOnFailureListener(c::resumeWithException) }
    private suspend fun awaitUpload(path: String, bytes: ByteArray, chargedBytes: Long) = suspendCancellableCoroutine<Unit> { c ->
        val metadata = com.google.firebase.storage.StorageMetadata.Builder().setCustomMetadata("organizationChargedBytes", chargedBytes.toString()).build()
        storage.reference.child(path).putBytes(bytes, metadata).addOnSuccessListener { c.resume(Unit) {} }.addOnFailureListener(c::resumeWithException)
    }
    private suspend fun awaitDownload(path: String) = suspendCancellableCoroutine<ByteArray> { c -> storage.reference.child(path).getBytes(250L * 1024 * 1024).addOnSuccessListener { c.resume(it) {} }.addOnFailureListener(c::resumeWithException) }
    private suspend fun awaitStorageDelete(path: String) = suspendCancellableCoroutine<Unit> { c -> storage.reference.child(path).delete().addOnSuccessListener { c.resume(Unit) {} }.addOnFailureListener(c::resumeWithException) }
    private suspend fun deleteTrackedStorage(context: Context, spaceId: String, path: String) {
        val reference = storage.reference.child(path)
        val charged = suspendCancellableCoroutine<Long> { c -> reference.metadata
            .addOnSuccessListener { c.resume(it.getCustomMetadata("organizationChargedBytes")?.toLongOrNull() ?: 0L) {} }
            .addOnFailureListener { c.resume(0L) {} } }
        awaitStorageDelete(path)
        com.arcinteractive.spaces.data.organization.OrganizationService().releaseStorageForSpace(context, spaceId, charged)
    }
}

data class NoteViewPreference(val lastViewedAt: Date, val viewCount: Int)

File: iOS/Spaces V1.5/Services/NoteService.swift
import CryptoKit
import FirebaseAuth
import FirebaseFirestore
import FirebaseStorage
import Foundation

@MainActor
final class NoteService {
    private let firestore = Firestore.firestore()
    private let storage = Storage.storage()
    private let encryption = EncryptionService()
    private let spaces = SpaceService()

    func listen(in space: Space, onUpdate: @escaping (Result<[SpaceNote], Error>) -> Void) -> ListenerRegistration {
        firestore.collection("spaces").document(space.id).collection("notes")
            .order(by: "updatedAt", descending: true).addSnapshotListener { [weak self] snapshot, error in
                if let error { onUpdate(.failure(error)); return }
                Task { @MainActor in
                    do {
                        let docs = snapshot?.documents ?? []
                        guard !docs.isEmpty else { onUpdate(.success([])); return }
                        guard let self else { return }
                        let key = try await self.key(space.id)
                        onUpdate(.success(try docs.map { try self.mapNote($0, spaceID: space.id, key: key) }))
                    } catch { onUpdate(.failure(error)) }
                }
            }
    }

    func listenToComments(space: Space, noteID: String, onUpdate: @escaping (Result<[SpaceNoteComment], Error>) -> Void) -> ListenerRegistration {
        firestore.collection("spaces").document(space.id).collection("notes").document(noteID).collection("comments")
            .order(by: "createdAt").addSnapshotListener { [weak self] snapshot, error in
                if let error { onUpdate(.failure(error)); return }
                Task { @MainActor in
                    do {
                        let docs = snapshot?.documents ?? []
                        guard !docs.isEmpty else { onUpdate(.success([])); return }
                        guard let self else { return }
                        let key = try await self.key(space.id)
                        let comments = try docs.map { doc -> SpaceNoteComment in
                            let data = doc.data()
                            let decrypted = try self.encryption.decryptData(ciphertext: data["ciphertext"] as? String ?? "", nonce: data["nonce"] as? String ?? "", using: key)
                            let payload = try JSONDecoder().decode(SpaceNoteCommentPayload.self, from: decrypted)
                            return .init(id: doc.documentID, noteID: noteID, authorID: data["authorId"] as? String ?? "", authorName: payload.authorName, body: payload.body, createdAt: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date())
                        }
                        onUpdate(.success(comments))
                    } catch { onUpdate(.failure(error)) }
                }
            }
    }

    func save(_ note: SpaceNote, in space: Space) async throws {
        guard let uid = Auth.auth().currentUser?.uid else { throw NoteError.notSignedIn }
        let isNew = note.createdBy.isEmpty
        let permission: SpacePermission = isNew ? .createNotes : (note.createdBy == uid ? .editOwnNotes : .editAnyNotes)
        guard await spaces.canPerform(permission, in: space) else { throw NoteError.permissionDenied }
        let encrypted = try encryption.encryptData(JSONEncoder().encode(SpaceNotePayload(title: note.title, markdown: note.markdown, attachments: note.attachments, links: note.links)), using: try await key(space.id))
        try await firestore.collection("spaces").document(space.id).collection("notes").document(note.id).setData([
            "ciphertext": encrypted.ciphertext, "nonce": encrypted.nonce,
            "createdBy": isNew ? uid : note.createdBy,
            "createdAt": isNew ? FieldValue.serverTimestamp() : Timestamp(date: note.createdAt),
            "updatedAt": FieldValue.serverTimestamp(), "encryptionVersion": "space-module-key-v1"
        ], merge: true)
        if isNew {
            await spaces.recordModuleActivity(
                type: .noteCreated,
                in: space,
                title: "created a Note",
                subtitle: note.title,
                targetID: note.id,
                targetType: .notes
            )
        }
        if !isNew {
            let mentionedIDs = await spaces.mentionedMemberIDs(in: note.markdown, space: space)
            await spaces.recordTargetedModuleNotification(
                recipientIDs: mentionedIDs,
                type: .note,
                in: space,
                title: "mentioned you in a Note",
                subtitle: note.title,
                targetID: note.id,
                targetType: .notes
            )
        }
    }

    func delete(_ note: SpaceNote, in space: Space) async throws {
        guard let uid = Auth.auth().currentUser?.uid else { throw NoteError.notSignedIn }
        guard await spaces.canPerform(note.createdBy == uid ? .deleteOwnNotes : .deleteAnyNotes, in: space) else { throw NoteError.permissionDenied }
        let ref = firestore.collection("spaces").document(space.id).collection("notes").document(note.id)
        for attachment in note.attachments { try? await storage.reference(withPath: attachment.storagePath).delete() }
        let comments = try await ref.collection("comments").getDocuments()
        for comment in comments.documents { try await comment.reference.delete() }
        try await ref.delete()
    }

    func addComment(_ body: String, authorName: String, note: SpaceNote, space: Space) async throws {
        guard let uid = Auth.auth().currentUser?.uid else { throw NoteError.notSignedIn }
        let encrypted = try encryption.encryptData(JSONEncoder().encode(SpaceNoteCommentPayload(authorName: authorName, body: body)), using: try await key(space.id))
        try await firestore.collection("spaces").document(space.id).collection("notes").document(note.id).collection("comments").document().setData([
            "ciphertext": encrypted.ciphertext, "nonce": encrypted.nonce, "authorId": uid,
            "createdAt": FieldValue.serverTimestamp(), "encryptionVersion": "space-module-key-v1"
        ])
        let mentionedIDs = await spaces.mentionedMemberIDs(in: body, space: space)
        await spaces.recordTargetedModuleNotification(
            recipientIDs: Array(Set(mentionedIDs + [note.createdBy])),
            type: .note,
            in: space,
            title: "commented on your Note",
            subtitle: note.title,
            targetID: note.id,
            targetType: .notes
        )
    }

    func recordView(noteID: String, spaceID: String) async {
        guard let uid = Auth.auth().currentUser?.uid else { return }
        try? await firestore.collection("spaces").document(spaceID).collection("notePreferences").document(uid).collection("notes").document(noteID).setData([
            "lastViewedAt": FieldValue.serverTimestamp(), "viewCount": FieldValue.increment(Int64(1))
        ], merge: true)
    }

    func listenToViewPreferences(
        spaceID: String,
        onUpdate: @escaping ([String: NoteViewPreference]) -> Void
    ) -> ListenerRegistration? {
        guard let uid = Auth.auth().currentUser?.uid else { return nil }
        return firestore.collection("spaces").document(spaceID)
            .collection("notePreferences").document(uid).collection("notes")
            .addSnapshotListener { snapshot, _ in
                let pairs: [(String, NoteViewPreference)] = (snapshot?.documents ?? []).compactMap { document -> (String, NoteViewPreference)? in
                    guard document.documentID != "_organization" else { return nil }
                    return (
                        document.documentID,
                        NoteViewPreference(
                            lastViewedAt: (document.data()["lastViewedAt"] as? Timestamp)?.dateValue() ?? .distantPast,
                            viewCount: document.data()["viewCount"] as? Int ?? 0
                        )
                    )
                }
                let values: [String: NoteViewPreference] = Dictionary(uniqueKeysWithValues: pairs)
                onUpdate(values)
            }
    }

    func listenToManualOrder(spaceID: String, onUpdate: @escaping ([String]) -> Void) -> ListenerRegistration? {
        guard let uid = Auth.auth().currentUser?.uid else { return nil }
        return firestore.collection("spaces").document(spaceID)
            .collection("notePreferences").document(uid)
            .addSnapshotListener { snapshot, _ in
                onUpdate(snapshot?.data()?["manualOrder"] as? [String] ?? [])
            }
    }

    func saveManualOrder(_ ids: [String], spaceID: String) async {
        guard let uid = Auth.auth().currentUser?.uid else { return }
        try? await firestore.collection("spaces").document(spaceID)
            .collection("notePreferences").document(uid)
            .setData(["manualOrder": ids, "updatedAt": FieldValue.serverTimestamp()], merge: true)
    }

    func upload(data: Data, name: String, mimeType: String, isMedia: Bool, noteID: String, space: Space) async throws -> SpaceNoteAttachment {
        let id = UUID().uuidString
        let encrypted = try encryption.encryptData(data, using: try await key(space.id))
        let path = "spaces/\(space.id)/notes/\(noteID)/attachments/\(id).enc"
        try await storage.reference(withPath: path).putDataAsync(Data(base64Encoded: encrypted.ciphertext) ?? Data())
        return .init(id: id, name: name, mimeType: mimeType, storagePath: path, nonce: encrypted.nonce, isMedia: isMedia)
    }

    func download(_ attachment: SpaceNoteAttachment, space: Space) async throws -> Data {
        let data = try await storage.reference(withPath: attachment.storagePath).data(maxSize: 250 * 1024 * 1024)
        return try encryption.decryptData(ciphertext: data.base64EncodedString(), nonce: attachment.nonce, using: try await key(space.id))
    }

    private func key(_ spaceID: String) async throws -> SymmetricKey { try await spaces.encryptionKeyForModuleData(spaceID: spaceID) }
    private func mapNote(_ doc: DocumentSnapshot, spaceID: String, key: SymmetricKey) throws -> SpaceNote {
        let data = doc.data() ?? [:]
        let decrypted = try encryption.decryptData(ciphertext: data["ciphertext"] as? String ?? "", nonce: data["nonce"] as? String ?? "", using: key)
        let payload = try JSONDecoder().decode(SpaceNotePayload.self, from: decrypted)
        return .init(id: doc.documentID, spaceID: spaceID, title: payload.title, markdown: payload.markdown, attachments: payload.attachments, links: payload.links, createdBy: data["createdBy"] as? String ?? "", createdAt: (data["createdAt"] as? Timestamp)?.dateValue() ?? Date(), updatedAt: (data["updatedAt"] as? Timestamp)?.dateValue() ?? Date())
    }
}

struct NoteViewPreference {
    let lastViewedAt: Date
    let viewCount: Int
}

enum NoteError: LocalizedError {
    case notSignedIn, permissionDenied
    var errorDescription: String? { self == .notSignedIn ? "Sign in to use Notes." : "You do not have permission to change this Note." }
}
